### PR TITLE
feat: sign with env and output certificate

### LINKED ISSUE #2657: support passing environments to "signs"
This is a follow-up issue for the https://github.com/google/ko/issues/491, dropping here so we don't forget!

---

**Is your feature request related to a problem? Please describe.**
Currently, we don't have the support for passing `env` variables to [signing pipeline](https://goreleaser.com/customization/sign/).

**Describe the solution you'd like**
Add a `KV` support for the `signs` [config structure](https://github.com/goreleaser/goreleaser/blob/9af17a49adde674d73c429daec7775d6156495a0/pkg/config/config.go#L590-L600). The [default value](https://github.com/goreleaser/goreleaser/blob/9af17a49adde674d73c429daec7775d6156495a0/internal/pipe/sign/sign.go#L31) should be set to _empty_.

```yaml
signs:
  env:
    - FOO: 1
    - BAR: "string"
    - BAZ: true
```

**Describe alternatives you've considered**

`-`

**Additional context**
I'm not sure whether we should look up the OS environment using the [expand()](https://github.com/goreleaser/goreleaser/blob/9af17a49adde674d73c429daec7775d6156495a0/internal/pipe/sign/sign.go#L194-L198) function. And not sure how logic should work here in case there is `$COSIGN_EXPERIMENTAL` in the environment. Would it be a bad practice to look at the _real_ OS environment instead of creating new `env` for _each_ pipe we want to set some envs laterly?

cc: @developer-guy @erkanzileli

### LINKED ISSUE #2659: Support Cosign keyless signing & verification
**Is your feature request related to a problem? Please describe.**
It is not related to a problem. It's a feature request for signing and verifying with Keyless Mode.[^1]

**Describe the solution you'd like**

GoReleaser can use the `cosign` tool as a command in the `sigs` section.  In order to enable Keyless mode in `cosign`, we need to pass an environment variable called `COSIGN_EXPERIMENTAL=1`. So, we've opened another issue to enable passing environment variables to the `sigs` section. Also, recently, the Kyverno project did the same thing. [^2]

[ ] - https://github.com/goreleaser/goreleaser/issues/2657 - @Dentrax 

**Describe alternatives you've considered**

**Additional context**
[^1]: https://github.com/sigstore/cosign/blob/main/KEYLESS.md
[^2]: github.com/kyverno/kyverno/pull/2677

cc: @dentrax @erkanzileli

### PR DESCRIPTION
**Pull request by caarlos0**
closes #2659
closes #2657

### CODE REVIEW

**Review by shibumi — APPROVED** [review]

**Comment by shibumi** [comment]
LGTM, good work @caarlos0 ! :)

**Review by developer-guy — APPROVED** [review]
LGTM, too 🙋🏻‍♂️🤙 thank you so much @caarlos0

File: pkg/config/config.go [hunk 1]
```diff
@@ -589,14 +589,16 @@ type NFPMOverridables struct {
...
+	ID          string   `yaml:"id,omitempty"`
+	Cmd         string   `yaml:"cmd,omitempty"`
+	Args        []string `yaml:"args,omitempty"`
+	Signature   string   `yaml:"signature,omitempty"`
+	Artifacts   string   `yaml:"artifacts,omitempty"`
+	IDs         []string `yaml:"ids,omitempty"`
+	Stdin       *string  `yaml:"stdin,omitempty"`
+	StdinFile   string   `yaml:"stdin_file,omitempty"`
+	Env         []string `yaml:"env,omitempty"`
+	Certificate string   `yaml:"pem,omitempty"`
```
Comment (developer-guy) [inline pkg/config/config.go:601]: should we change it to `certificate` instead of `pem`? @caarlos0 @shibumi ?

Comment (caarlos0) [reply pkg/config/config.go:601]: Yes, bad rename on my part 😂

Good catch!

File: internal/exec/exec.go [hunk 2]
```diff
@@ -114,7 +114,7 @@ func filterArtifacts(artifacts artifact.Artifacts, publisher config.Publisher) [
 	}
 
 	if publisher.Signature {
-		filters = append(filters, artifact.ByType(artifact.Signature))
+		filters = append(filters, artifact.ByType(artifact.Signature), artifact.ByType(artifact.Certificate))
```
Comment (caarlos0) [inline internal/exec/exec.go:117]: todo: add tests for this

File: internal/http/http.go [hunk 3]
```diff
@@ -149,7 +149,7 @@ func Upload(ctx *context.Context, uploads []config.Upload, kind string, check Re
 			filters = append(filters, artifact.ByType(artifact.Checksum))
 		}
 		if upload.Signature {
-			filters = append(filters, artifact.ByType(artifact.Signature))
+			filters = append(filters, artifact.ByType(artifact.Signature), artifact.ByType(artifact.Certificate))
```
Comment (caarlos0) [inline internal/http/http.go:152]: todo: add tests for this

File: internal/pipe/blob/upload.go [hunk 4]
```diff
@@ -81,6 +81,7 @@ func doUpload(ctx *context.Context, conf config.Blob) error {
 		artifact.ByType(artifact.UploadableSourceArchive),
 		artifact.ByType(artifact.Checksum),
 		artifact.ByType(artifact.Signature),
+		artifact.ByType(artifact.Certificate),
```
Comment (caarlos0) [inline internal/pipe/blob/upload.go:84]: todo: add tests for this

**Comment by caarlos0** [comment]
related to this PR: #2680

**Pull request merged**